### ISSUE TITLE
After update issue ERR_BLOCKED_BY_RESPONSE

### ISSUE THREAD

**Issue by krishnaTORQUE** [helmetjs/helmet#343]
I am getting this error after update
`net::ERR_BLOCKED_BY_RESPONSE.NotSameOriginAfterDefaultedToSameOriginByCoep 200`

**Comment by EvanHahn** [comment]
Sorry about this.

1. Where is the error coming from?
2. Could you create a small sample app that demonstrates the problem?

**Comment by Benno007** [comment]
Same issue with us but we just didn't know the changes in the changelog would affect us.

We just have a web site with a script tag that loads a javascript file from an API endpoint.
Example:
app.example.com has: `<script src="https://some-api.example.com"></script>` which loads javascript. After upgrading to helmet 5, we are now getting this error `Failed to load resource: net::ERR_BLOCKED_BY_RESPONSE.NotSameOrigin`

We are just doing this in our express app though and it worked immediately before this PR went in. We have seen in your changelog that the defaults have changed for the cross origin resource policy. We are in the middle of a release now so just reverting helmet, but will look at updating it with the correct options.

```
// Our some-api express server has this:
    this.express.use(helmet())
```

**Comment by EvanHahn** [comment]
In the immediate term, downgrading to Helmet v4 will solve your problem.

A small sample app demonstrating the issue would be super useful.

**Comment by krishnaTORQUE** [comment]
The is showing in Devtools
`net::ERR_BLOCKED_BY_RESPONSE.NotSameOriginAfterDefaultedToSameOriginByCoep 200`

Helmet Config
```javascript
app.use(
  helmet({
    contentSecurityPolicy: {
      directives: {
        defaultSrc: ["'self'", 'google.com', 'youtube.com'],
        styleSrc: ["'self'", 'google.com', 'youtube.com', "'unsafe-inline'"],
        scriptSrc: ["'self'", 'google.com', 'youtube.com', "'unsafe-inline'"],
        imgSrc: ["'self'", 'google.com', 'youtube.com', 'data:', 'blob:'],
        connectSrc: ["'self'", 'google.com', 'youtube.com', 'ws:', 'wss:'],
        frameSrc: ["'self'", 'google.com', 'youtube.com'],
        frameAncestors: ["'self'", 'google.com', 'youtube.com'],
        mediaSrc: ["'self'", 'google.com', 'youtube.com', 'data:', 'blob:'],
      },
    },
    frameguard: { action: 'SAMEORIGIN' },
  })
);

```

This same code works in `v4.x` but not in `v5.x`

**Comment by EvanHahn** [comment]
What resource are you trying to load?

Could you create a sample app that demonstrates the issue? Maybe with codesandbox.io?

**Comment by krishnaTORQUE** [comment]
When loading `.js` content but this is not quite easy to replicate.
Its working fine in DEV but when in `Production` specially `example.com` try to load `static.example.com` its giving the error.

I guess @Benno007 has the same issue on subdomain.

**Comment by EvanHahn** [comment]
Sorry about this.

Once we've got a reliable reproduction, I'll take a look.

**Comment by manuelpoelzl** [comment]
Hi!
After I am also affected by the error and read along here, I quickly created an app on codesandbox, which shows the problem
https://codesandbox.io/s/elegant-sinoussi-v9o06?file=/app.js

I am trying to generate an avatar using https://eu.ui-avatars.com/api/?name=John+Doe , and this is exactly where this error occurs, as @krishnaTORQUE  already describes, it also occurs here with the subdomain.

Attached is also a screenshot showing the error

![Screenshot_1](https://user-images.githubusercontent.com/46539758/148204320-f0b7ee34-efcd-48b9-a4e4-cb0b36547961.png)

**Comment by EvanHahn** [comment]
@manuelpoelzl Thanks for this. I'll take a look later today.

**Comment by Monokai** [comment]
Same issue with loading an external Stripe checkout script. Works locally, on production it fails. Downgrading to 4.6 solves it.

**Comment by EvanHahn** [comment]
**tl;dr: disable [the `Cross-Origin-Embedder-Policy` header][0], enabled by default in Helmet v5.**

```js
app.use(
  helmet({
    crossOriginEmbedderPolicy: false,
    // ...
  })
);
```

---

Helmet v5 sets the [the `Cross-Origin-Embedder-Policy` HTTP response header][0] to `require-corp`. (This was possible in Helmet v4, but it was off by default, so most people didn't use it.)

Setting this header means that loading cross-origin resources (like an image from another resource) is trickier. For example, loading a cross-origin like this...

```html
<img alt="My picture" src="https://example.com/image.png">
```

...won't work unless `example.com` explicitly allows it, by setting some response headers of its own. Your browser will try to load `example.com/image.png`, and if it's not explicitly allowed, your browser will drop the response.

To fix this, you can prevent Helmet from setting the `Cross-Origin-Embedder-Policy` header, like this:

```js
app.use(
  helmet({
    crossOriginEmbedderPolicy: false,
    // ...
  })
);
```

I made [a small sample app](https://codesandbox.io/s/cold-rain-flk19?file=/src/index.js) you can use to play around with this. In my testing, it doesn't seem to work in HTTP but it _does_ over HTTPS, which might explain why things only break in production.

I _think_ this solves the problem so I'm going to close this issue. Let me know if that's wrong and I'll reopen!

[0]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cross-Origin-Embedder-Policy

**Comment by krishnaTORQUE** [comment]
@EvanHahn thanks, it worked

**Comment by K-404** [comment]
@EvanHahn Hi, Evan,
have the same issue here. Is it possible to turn it off standalone? 
something like: app.use(helmet.crossOriginEmbedderPolicy(false));

**Comment by EvanHahn** [comment]
@K-404 Not quite. Your options are:

1. Use the top-level Helmet and disable the `crossOriginEmbedderPolicy` middleware.

   ```js
   app.use(
     helmet({
       crossOriginEmbedderPolicy: false,
       // ...
     })
   );
   ```
2. I do not recommend this, but you can use the top-level Helmet and then manually remove the `Cross-Origin-Embedder-Policy` header later.

   ```js
   app.use(helmet());
   
   app.use((req, res, next) => {
     res.removeHeader("Cross-Origin-Embedder-Policy");
     next();
   });
   ```
   
   This should work, but it couples two middlewares. Why not just skip it at the top level?
3. I do not recommend this, but you can break Helmet into its smaller middlewares and skip `helmet.crossOriginEmbedderPolicy`.

   ```diff
    app.use(helmet.contentSecurityPolicy());
   -app.use(helmet.crossOriginEmbedderPolicy());
    app.use(helmet.crossOriginOpenerPolicy());
    app.use(helmet.crossOriginResourcePolicy());
    app.use(helmet.dnsPrefetchControl());
    app.use(helmet.expectCt());
    app.use(helmet.frameguard());
    app.use(helmet.hidePoweredBy());
    app.use(helmet.hsts());
    app.use(helmet.ieNoOpen());
    app.use(helmet.noSniff());
    app.use(helmet.originAgentCluster());
    app.use(helmet.permittedCrossDomainPolicies());
    app.use(helmet.referrerPolicy());
    app.use(helmet.xssFilter());
   ```

   This should work, but it will make Helmet harder to upgrade in the future if middlewares are added or removed.

Hope this helps!

**Comment by bluedge** [comment]
Having the same issue with polyfill.io. I ended up doing this (works)
Is there a more elegant solution?

```
// HELMET Content Security Policy
app.use(
	helmet.contentSecurityPolicy({
		useDefaults: true,
		directives: {
			'img-src': ["'self' data:", '*.google-analytics.com', '*.vimeocdn.com'],
			'script-src': ["'self'", '*.polyfill.io', "'unsafe-eval'", "'unsafe-inline'"],
			'default-src': ["'self'", '*.google-analytics.com', '*.gstatic.com', '*.googleapis.com', 'vimeo.com', '*.vimeo.com']
		}
	})
);
app.use((req, res, next) => {
	res.removeHeader("Cross-Origin-Embedder-Policy");
	next();
});
```

**Comment by EvanHahn** [comment]
@bluedge That should work but I think it'd be better to just prevent Helmet from setting the header in the first place. See my previous comment where I show how to do this. Email me@evanhahn.com if you need help.

**Comment by bluedge** [comment]
@EvanHahn I tried other solutions based on your previous comments bit it would still trigger the error for `*.polyfill.io`
`ERR_BLOCKED_BY_RESPONSE.NotSameOriginAfterDefaultedToSameOriginByCoep 200`

**Comment by EvanHahn** [comment]
@bluedge Could you include a code snippet that reproduces the problem for you?

**Comment by amitagrawal11** [comment]
Got the same issue, after looking into each ticket I got to know the ticket is already raised. 

It would be helpful for other forks if you can change the title of the issue to "net::ERR_BLOCKED_BY_RESPONSE.NotSameOriginAfterDefaultedToSameOriginByCoep" so that anybody can search it on google too.

**Comment by EvanHahn** [comment]
@amitagrawal11 Good idea. I'll consider creating a separate resource for this.

**Comment by EvanHahn** [comment]
@amitagrawal11 I published this [as a Stack Overflow question and answer](https://stackoverflow.com/q/71904052/804100), which I think will help more people as they search online for the solution. Feel free to make a post on your own blog copying my answer if you wish!

**Comment by CoderMian** [comment]
// https://nuxt.com/docs/api/configuration/nuxt-config
export default defineNuxtConfig({
  modules: [
    "@nuxt/image",
    "nuxt-swiper",
    "@pinia/nuxt",
    "@pinia-plugin-persistedstate/nuxt",
  "nuxt-security",
  ],
  pinia: {
    storesDirs: ["./stores/**"],
  },
  image: {
    format: ["webp", "jpg", "jpeg", "png", "svg", "gif", "avif"],
  },
  css: ["~/assets/css/style.css"],

  

  routeRules: {
    "/checkout": { ssr: false },
    "/signup": { ssr: false },
  },
  security: {
    headers: {
      crossOriginResourcePolicy: "cross-origin",
      crossOriginOpenerPolicy: "unsafe-none",
      crossOriginEmbedderPolicy: "require-corp",
      contentSecurityPolicy: {
        
        "img-src": [
          "'self'",
          "data:",
          "https://cdn.sellfront.co",
          "https://raw.githubusercontent.com",
          "https://meriwebshop.sgp1.digitaloceanspaces.com/",
          "https://connect.facebook.net",
        ],
        "media-src": [
          "'self'",
          "https://cdn.sellfront.co",
          "https://meriwebshop.sgp1.digitaloceanspaces.com/",
        ],
        "worker-src": ["'none'"],
        "style-src": [
          "'self'",
          "report-sample",
          "https://fonts.googleapis.com",
          "https://cdn.jsdelivr.net",
          "'unsafe-inline'",
        ],
        "script-src": [
          "'self'",
          "'unsafe-inline'",
          "http:",
          "https://cdn.jsdelivr.net",
          "https://cdnjs.cloudflare.com",
          "http://localhost:3002/",
          "http://localhost:3002/_nuxt/",
          "https://www.googletagmanager.com/",
          "https://connect.facebook.net",
        ],
        "script-src-elem": [
          "'self'",
          "'unsafe-inline'",
          "unsafe-hashes",
          "https://cdn.jsdelivr.net",
          "https://cdnjs.cloudflare.com",
          "http://localhost:3002/",
          "http://localhost:3002/_nuxt/",
          "https://www.googletagmanager.com/",
          "https://connect.facebook.net",
        ],
        "font-src": [
          "'self'",
          "data:",
          "https://fonts.googleapis.com",
          "https://fonts.gstatic.com",
        ],
        "frame-src": ["'self'"],
        "manifest-src": ["'self'"],
        "connect-src": [
          "'self'",
          process.env.BASE_URL || "",
          "http://localhost:3002/",
        ],
        "form-action": ["'self'"],
        "frame-ancestors": ["'none'"],
        "object-src": ["'none'"],
        "base-uri": ["'self'"],
        "child-src": ["'self'", "https://www.google.com"],
      },
    },
  },
});

This is my nuxt.config.ts file and I am using nuxt 3 I want to implement security headers but after install nuxt-security and add in nuxt,config.ts I am setting my headers but got this error I dont understand error please help me to fix this



![image](https://github.com/helmetjs/helmet/assets/104327746/f81e8450-a9e0-47ca-ab06-aad8b9fc7e5b)

**Comment by glensc** [comment]
@CoderMian put your post properly between code fence (edit your post):
- https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks

**Comment by EvanHahn** [comment]
Please open an issue on [nuxt-security](https://github.com/Baroshem/nuxt-security).

**Comment by EvanHahn** [comment]
@CoderMian Please open an issue on [nuxt-security](https://github.com/Baroshem/nuxt-security).

**Comment by deepkuma123** [comment]
import express from "express";
import bodyParser from "body-parser";
import mongoose from "mongoose";
import cors from "cors";
import dotenv from "dotenv";
import multer from "multer";
import helmet from "helmet";
import morgan from "morgan";
import path from "path";
import { fileURLToPath } from "url";
import authRoutes from "./routes/auth.js";
import userRoutes from "./routes/users.js";
import postRoutes from "./routes/posts.js";
import { register } from "./controllers/auth.js";
import { createPost } from "./controllers/posts.js";
import { verifyToken } from "./middleware/auth.js";
import User from "./models/User.js";
import Post from "./models/Post.js";
import { users, posts } from "./data/index.js";
import http from "http";
import { Server } from "socket.io";
import { existsSync } from "fs";

/* CONFIGURATIONS */
const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
dotenv.config();
const app = express();

app.use(express.json());
app.use(helmet());
app.use(helmet.crossOriginResourcePolicy({ policy: "cross-origin" }));
app.use(morgan("common"));
app.use(bodyParser.json({ limit: "30mb", extended: true }));
app.use(bodyParser.urlencoded({ limit: "30mb", extended: true }));
app.use(cors());
app.use("/assets", express.static(path.join(__dirname, "/public/assets")));

app.use((req, res, next) => {
  res.setHeader(
    "Content-Security-Policy",
    "img-src 'self' data: https://firebasestorage.googleapis.com"
  );
  next();
});
// app.use((req, res, next) => {
//   res.setHeader("Cross-Origin-Resource-Policy", "cross-origin");
//   next();
// });
console.log(__dirname);
/* FILE STORAGE */
const storage = multer.diskStorage({
  destination: function (req, file, cb) {
    console.log("this one");
    cb(null, "server/public/assets");
  },
  
});
const upload = multer({ storage });

/* ROUTES WITH FILES */
app.post("/auth/register", upload.single("picture"), register);
app.post("/posts", verifyToken, upload.single("picture"), createPost);

/* ROUTES */
app.use("/auth", authRoutes);
app.use("/users", userRoutes);
app.use("/posts", postRoutes);

/* Serve the frontend files */
// Serve the frontend files
const frontendPath = path.join(__dirname, "../client", "build");
console.log(frontendPath);
if (existsSync(frontendPath)) {
  // Only serve frontend files if the directory exists
  app.use(express.static(frontendPath));
  app.get("*", (req, res) => {
    res.sendFile(path.join(frontendPath, "index.html"));
  });
}

/* MONGOOSE SETUP */
const PORT = process.env.PORT || 6001;
/* ADD DATA ONE TIME */
// User.insertMany(users);
// Post.insertMany(posts);
const httpServer = http.createServer(app); // Create an HTTP server
const io = new Server(httpServer, {
  cors: {
    origin: [
      "http://localhost:3000",
      "http://192.168.99.124:3000",
      "https://firebasestorage.googleapis.com",
    ],
  },
});

// Socket.io connection handling
io.on("connection", (socket) => {
  console.log("A user connected");

  // Handle new post event
  socket.on("new_post", (postData) => {
    console.log("New post received:", postData);
    // Broadcast the new post to all connected clients
    io.emit("new_post", postData);
  });

  // Handle "like_patched" event
  socket.on("like_patched", (postId, userId) => {
    // Perform necessary actions here, such as updating the database with the patched like
    console.log(postId);
    // Emit the updated post to all connected clients
    io.emit("post_updated", postId);
  });

  // Handle delete post event
  socket.on("delete_post", (postId) => {
    console.log("Post deleted:", postId);
    // Broadcast the deleted post ID to all connected clients
    io.emit("delete_post", postId);
  });

  socket.on("comment_created", (newComment) => {
    console.log(newComment);
    io.emit("comment_created", newComment);
  });
  socket.on("delete_comment", (response) => {
    console.log(response);
    // Then, emit an event to notify clients that the comment has been deleted
    io.emit("delete_comment", response);
  });

  // Add other event handlers as needed
  // Handle disconnect event
  socket.on("disconnect", () => {
    console.log("A user disconnected");
  });
});

// Mongoose setup and server listening
mongoose
  .connect(process.env.MONGO_URL, {
    useNewUrlParser: true,
    useUnifiedTopology: true,
  })
  .then(() => {
    httpServer.listen(PORT, () => {
      console.log(`Server Port: ${PORT}`);
    });

    // Additional setup code if needed
  })
  .catch((error) => console.log(`${error} did not connect`));
  
  
  ### this is my server code and getting the error 
  
  GET https://firebasestorage.googleapis.com/v0/b/social-storage-d8947.appspot.com/o/images%2F1.jpg?alt=media&token=16663329-aa89-4a64-a205-4bfe9cc213ef net::ERR_BLOCKED_BY_RESPONSE.NotSameOriginAfterDefaultedToSameOriginByCoep 200 (OK)
  
  can someone help me to solve this issue

**Comment by EvanHahn** [comment]
@deepkuma123 Please open a new issue. (It looks like you might be on an old version of Helmet?)

**Comment by MustafaGhouri** [comment]
`app.use(helmet.crossOriginResourcePolicy({ policy: "cross-origin" }));`
this was works for me
